nano/node/request_aggregator.hpp

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <nano/lib/locks.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/node/transport/transport.hpp>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+#include <condition_variable>
+#include <thread>
+#include <unordered_map>
+
+namespace mi = boost::multi_index;
+
+namespace nano
+{
+class votes_cache;
+class block_store;
+class wallets;
+class networ_params;
+class stat;
+/**
+ * Pools together confirmation requests, separately for each endpoint.
+ * Requests are added from network messages, and aggregated to minimize bandwidth and vote generation. Example:
+ * * Two votes are cached, one for hashes {1,2,3} and another for hashes {4,5,6}
+ * * A request arrives for hashes {1,4,5}. Another request arrives soon afterwards for hashes {2,3,6}
+ * * The aggregator will reply with the two cached votes
+ * Votes are generated for uncached hashes.
+ */
+class request_aggregator final
+{
+	/**
+	 * Holds a buffer of incoming requests from an endpoint.
+	 * Extends the lifetime of the corresponding channel. The channel is updated on a new request arriving from the same endpoint, such that only the newest channel is extended
+	 */
+	struct channel_pool final
+	{
+		channel_pool () = delete;
+		explicit channel_pool (std::shared_ptr<nano::transport::channel> & channel_a) :
+		start (std::chrono::steady_clock::now ()),
+		channel (channel_a),
+		endpoint (nano::transport::map_endpoint_to_v6 (channel_a->get_endpoint ()))
+		{
+		}
+		std::vector<std::pair<nano::block_hash, nano::root>> hashes_roots;
+		std::chrono::steady_clock::time_point start;
+		std::chrono::steady_clock::time_point deadline;
+		nano::endpoint endpoint;
+		std::shared_ptr<nano::transport::channel> channel;
```

Comment:
channel and endpoint should be initialized in the order of the class members. You could also initialize some of them in-class.